okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java

```diff
@@ -69,21 +74,27 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
     spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
 
     Platform platform = Platform.get();
-    try {
-      // Force handshake.
-      sslSocket.startHandshake();
-
-      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-      if (negotiatedProtocol == null) {
-        throw new RuntimeException("protocol negotiation failed");
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
+    // when creates the SSLSocket.
+    String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
+    if (negotiatedProtocol == null) {
+
+      try {
+        // Force handshake.
+        sslSocket.startHandshake();
+
+        negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
+        if (negotiatedProtocol == null) {
+          throw new RuntimeException("protocol negotiation failed");
+        }
+      } finally {
+        platform.afterHandshake(sslSocket);
```

Comment:
Hmm, seems it got changed, see https://github.com/square/okhttp/blob/parent-2.2.0/okhttp/src/main/java/com/squareup/okhttp/ConnectionSpec.java#L143